Adding Documentation for ECMA-402 (#497)
facebook/hermes: doc/intl.md

> # Overview

Reviewer: Before this, can you add an ID and title for docusaurus?
The "id" is used in hermes/website/sidebars.json as a key, and "title" is displayed as the overarching title.

Make sure to also add to the sidebars.json file so this will be displayed on our website.
```suggestion
---
id: intl
title: Intl Library
---
# Overview
```

Reviewer 2: maybe `title: Internationalization APIs`

> This document describes the current state of Android implementation of Intl APIs(formally ECMA-402 specification) in Hermes JavaScript engine. ECMA-402 is still evolving and the latest iteration is [7th edition](https://402.ecma-international.org/7.0/) which was published in June 2020. Each new edition is built on top of the last one and typically adds new capabilities typically as,

Reviewer 2: ```suggestion
This document describes the current state of Android implementation of [the ECMAScript Internationalization API Specification](https://tc39.es/ecma402/) (ECMA-402, or `Intl`) in Hermes JavaScript engine. ECMA-402 is still evolving and the latest iteration is [7th edition](https://402.ecma-international.org/7.0/) which was published in June 2020. Each new edition is built on top of the last one and typically adds new capabilities typically as,

```

We should avoid the "Intl". Either go with `Intl` (note that it's formatted as code) or "Internationalization".

> - New Intl service constructors (for e.g. Intl.Collator, Intl.NumberFormat etc...) or extending existing ones by accepting more parameters

Reviewer 2: ```suggestion
- New `Intl` service constructors (for e.g. `Intl.Collator`, `Intl.NumberFormat`, etc.) or extending existing ones by accepting more parameters.

```

Same to other places XD

> - New functions or properties in the Intl object prototypes (for e.g. Intl.Collator.prototype.compare)

Reviewer 2: ```suggestion
- New functions or properties in the `Intl` object prototypes (e.g. `Intl.Collator.prototype.compare`)
```

- New locale aware functions in standard Javascript object prototypes (for e.g. String.prototype.localeCompare)

> One standard internationalization implementation strategy followed by other engines, is to bundle a library and data (typically [icu](http://site.icu-project.org/)) along with the application package. This guarantees deterministic functionality at the cost of applications package bloat.  We decided to consume the Android platform libraries and data for space efficiency, but at the cost of some variance in behaviour across  Android platforms.

Reviewer 2: ```suggestion
One standard internationalization implementation strategy followed by other engines, is to bundle a library and data (typically [ICU](http://site.icu-project.org/)) along with the application package. This guarantees deterministic functionality at the cost of applications package bloat. We decided to consume the Android platform libraries and data for space efficiency, but at the cost of some variance in behaviors across Android platforms.

```

> # ECMA-402 Compliance

Reviewer 2: ❤️

Author: Ah .. I thought about your question last week .. but missed actually replying to it ! 

I am hoping we will continue to keep the docx as a richer detailed documentation for the Intl developers and the markdown as the quick reference for Intl consumers .. I see that many aspects of our representations in the docx can be useful for the consumers too, but will also also make the contents complicated .. I think we could work on it as part of our next wave of Intl feature developement ..

## Supported

> - Intl.Collator

Reviewer 2: Please make sure those are backticks-quoted as well

  - Intl.DateTimeFormat.prototype.formatToParts
  - Intl.DateTimeFormat.prototype.resolvedOptions

- Intl.getCanonicalLocales

- String.prototype
  - localeCompare
  - toLocaleLowerCase
  - toLocaleUpperCase

- Array.prototype
  - toLocaleString

- Number.prototype
  - toLocaleString

- Date.prototype
  - toLocaleString
  - toLocaleDateString
  - toLocaleTimeString

## Not yet supported

- [Intl.PluralRules](https://tc39.es/ecma402/#pluralrules-objects)

- [Intl.RelativeTimeFormat](https://tc39.es/ecma402/#relativetimeformat-objects)

- [Intl.DisplayNames](https://tc39.es/proposal-intl-displaynames/#sec-intl-displaynames-constructor)

- [Intl.ListFormat](https://tc39.es/proposal-intl-list-format/#sec-intl-listformat-constructor)

- [Intl.Locale](https://tc39.es/ecma402/#sec-intl-locale-constructor)

- Intl.DateTimeFormat properties
   - [dateStyle/timeStyle](https://tc39.es/proposal-intl-datetime-style/)
   - [dayPeriod](https://github.com/tc39/ecma402/issues/29)
   - [fractionalSecondDigits](https://github.com/tc39/ecma402/pull/347)
- [BigInt.prototype.toLocaleString](https://tc39.es/ecma402/#sup-bigint.prototype.tolocalestring)

## Excluded

> - [formatMatcher](https://tc39.es/ecma402/#sec-basicformatmatcher) property in Intl.DateTimeFormat.

Reviewer 2: Could you explain with more details like what you did in `*.docx`? You also mentioned that Firefox/Chrome does not seem to respect this.

## Known Issues

### Android 11

> - The keys of the object returned by &#39;resolvedOptions&#39; function in all Intl services are not deterministically ordered as prescribed by spec.

Reviewer: Fix the HTML escape here, it should probably use backticks directly

> - DateFormat: ECMAScript [beginning of time](https://www.ecma-international.org/ecma-262/11.0/index.html#sec-time-values-and-time-range) (-8,640,000,000,000,000), is formatted as November 271817, instead of expected April 271822.

Reviewer 2: It's very important to properly format this sentence :) I'm assuming `April 271822` is the function output? They should be `-quoted

- Intl.NumberFormat implementation has some rough edges in supporting the following properties,
  - style: 'unit'
  - notation: 'compact'
  - signDisplay
  - currencyFormat

> ### Android 10 (SDK<30) and older

Reviewer 2: ```suggestion
### Android 10 and older (SDK < 30)

```

- NumberFormat: Scientific notation formatting has issues on some cases such as Infinity (e.g. Expected SameValue(«-∞E0», «-∞») to be true).

> - NumberFormat: Compact notation formatToParts doesn&#39;t identify unit, hence we report unit as &#39;literal&#39; (Compact short: 987654321: parts[1].type Expected SameValue(«literal», «compact») to be true)

Reviewer: Fixup the HTML escape here as well

### Android 8.0 – 8.1 (SDK<28) and older

- getCanonicalLocales: Some differences in the keyword values due to CLDR/Unicode version difference (Expected SameValue(«und-u-tz-utc», «und-u-tz-gmt») to be true)
- NumberFormat: CompactFormatter doesn&#39;t respect the precision inputs (Expected SameValue(«9,900만», «9877만») to be true; Expected SameValue(«990M», «988M») to be true).

### Android 7.0 - 7.1 (SDK<26) and older

- getCanonicalLocales: Unicode/CLDR version differences Expected SameValue(«und-u-ms-imperial», «und-u-ms-uksystem») to be true

### Android 7.0 - 7.1 (SDK<24) and older

- Collator: Doesn&#39;t canonically decompose the input strings. Canonically equivalent string with non-identical code points may not match.
- getCanonicalLocales: Unicode/CLDR version differences (Expected SameValue(«und-u-ca-ethiopic-amete-alem», «und-u-ca-ethioaa») to be true; Expected SameValue(«und-u-ks-primary», «und-u-ks-level1») to be true)
- Unit style does not work.
- There are issues in the precision configuration due to lack of APIs.
- DateFormat: There are issues with the calendar configuration which needs to be dug into.

> ### SDK<21 and older

Reviewer 2: ```suggestion
### SDK < 21
```

`<` says all

On platforms before 21, Locale.forLanguageTag() is not available, hence we can&#39;t construct java.util.Locale object from locale tag. Hence, we fallback to English for any locale input.

# Internationalization framework in Android Platform

Our implementation is essentially projecting the Android platform provided internationalization facilities through the ECMA-402 specified services. It implies that the results of running the same code may vary between devices running different versions of Android.

Android platform internationalization libraries have been based on [ICU4j project](https://unicode-org.github.io/icu-docs/#/icu4j). Version of ICU4j and the backing [CLDR data](http://cldr.unicode.org/) varies across Android platform versions. Also, the ICU APIs were never exposed directly, but only through wrappers and aliases. This results in significant variance in internationalization API surface and data across platform versions.

The following table summarizes ICU, CLDR and Unicode versions available on the Android platforms that we intent to support.

> **Platform 24+ where ICU4j APIs are available.**

Reviewer 2: LOVE THIS TABLE

| Android Platform Version | ICU | Unicode | CLDR
| --- | --- | --- | --- |
| Android 11 (API level 30) | ICU4J 66.1 ([ref](https://android.googlesource.com/platform/external/icu/+/refs/heads/android11-mainline-release/icu4j/readme.html)) | Unicode 13 beta | CLDR 36.1 |
| Android 10 (API level 29) | ICU4j 63.2 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 34 | Unicode 11.0 |
| Android 9 (API level 28) | ICU4j 60.2 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 32.0.1 | Unicode 10.0 |
| Android 8.0 - 8.1 (API levels 26 - 27) | ICU4j 58.2( [ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 30.0.3 | Unicode 9.0 |
| Android 7.0 - 7.1 (API levels 24 - 25) | ICU4j 56 ([ref](https://developer.android.com/guide/topics/resources/internationalization))| CLDR 28 | Unicode 8.0 |

> <br />
> 
> **Pre-24 platforms**

Reviewer 2: ```suggestion
### Pre-24 platforms
```

🤔   I haven't tried it locally but I imagined you can gain some proper margin by making it a header? (Same applies to the first table head)

| Android Platform Version | ICU | Unicode | CLDR
| --- | --- | --- | --- |
| Android 6.0 (API level 23) | ICU4j 55.1 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 27.0.1 | Unicode 7.0 |
| Android 5.0 (API levels 21–22) | ICU4j 53 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 25 | Unicode 6.3 |
| Android 4.4 (API levels 19–20) | ICU4j 51 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 23 | Unicode 6.2 |
| Android 4.3 (API level 18) | ICU4j 50 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 22.1 | Unicode 6.2 |
| Android 4.1 (API levels 16–17) | ICU4j 4.8 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 2.0 | Unicode 6.0 |  
  

> <br />

Reviewer 2: 😅

To Summarize, there are three big partitions in the platform domain,

Platforms <= 24 has much better internationalization support than earlier, as many ICU classes are available as is.

Platforms 21-24 still has reasonable internationalization support, by allowing creation of Locale objects and exposing selected ICU services through java.text namespace.

Platforms < 21 doesn't allow creation of Locale objects from tags, severely limiting general purpose international code.

> # Impact on Application Size

Reviewer 2: Yeah we should include this 👍

The following numbers are measured using a test application which takes dependency on the Hermes library to evaluate a JavaScript snippet. Essentially, enabling Intl APIs adds 57-62K per ABI.

> | **Product APK Size** | **NOINTL** | **INTL** | **DIFF** | **PERC** |

Reviewer 2: Awesome table!

## Limitations across Android SDKs

### Android 11

- The keys of the object returned by `resolvedOptions` function in all `Intl` services are not deterministically ordered as prescribed by spec.
- DateFormat: ECMAScript [beginning of time](https://www.ecma-international.org/ecma-262/11.0/index.html#sec-time-values-and-time-range) (-8,640,000,000,000,000), is formatted as `November 271817`, instead of expected `April 271822`.
- `Intl.NumberFormat` implementation has some rough edges in supporting the following properties,
  - `style`: 'unit'
  - `notation`: 'compact'
  - `signDisplay`
  - `currencyFormat`

### Android 10 and older (SDK < 30)

- `Intl.NumberFormat`: Scientific notation formatting has issues on some cases. e.g. `-Infinity` may get formatted as '-∞E0' instead of expected '-∞'. Another manifestation of the issues is that the formatToParts may return 4 parts instead of 2.

> - `Intl.NumberFormat`: Compact notation `formatToParts` doesn't identify unit, hence we report unit as 'literal'. For e.g. the second part of "100ac" gets reported as "literal" instead of "compact"

Reviewer 2: I really appreciate this example. Love it!

### Android 9 and older (SDK < 29)

- There are some failures likely due to older Unicode and CLDR version, which are hard to generalize. Some examples are,
  - `Intl.NumberFormat`: 'Percent' is not accepted as a unit.
  - `Intl.NumberFormat`: unit symbols difference, kph vs km/h
  - Some issue in significant digit precision, which is not yet looked into the details.

### Android 8.0 – 8.1 and older (SDK < 28)

- `Intl.getCanonicalLocales`: Some differences in the keyword values due to CLDR/Unicode version difference (Expected SameValue(«und-u-tz-utc», «und-u-tz-gmt») to be true)
- `Intl.NumberFormat`: CompactFormatter doesn't respect the precision inputs (Expected SameValue(«9,900만», «9877만») to be true; Expected SameValue(«990M», «988M») to be true).

### Android 7.0 - 7.1 and older (SDK < 26)

- `Intl.getCanonicalLocales`: Unicode/CLDR version differences Expected SameValue(«und-u-ms-imperial», «und-u-ms-uksystem») to be true

### Android 7.0 - 7.1 and older (SDK < 24)

- `Intl.Collator`: Doesn't canonically decompose the input strings. Canonically equivalent string with non-identical code points may not match.

> - `Intl.getCanonicalLocales`: Unicode/CLDR version differences (Expected SameValue(«und-u-ca-ethiopic-amete-alem», «und-u-ca-ethioaa») to be true; Expected SameValue(«und-u-ks-primary», «und-u-ks-level1») to be true)

Reviewer 2: There are still some `Expected SameValue` here that might confuse our reader

---
id: intl
title: Internationalization APIs
---

This document describes the current state of Android implementation of the [ECMAScript Internationalization API Specification](https://tc39.es/ecma402/) (ECMA-402, or `Intl`). ECMA-402 is still evolving and the latest iteration is [7th edition](https://402.ecma-international.org/7.0/) which was published in June 2020. Each new edition is built on top of the last one and adds new capabilities typically as,
- New `Intl` service constructors (e.g. `Intl.Collator`, `Intl.NumberFormat` etc.) or extending existing ones by accepting more parameters
- New functions or properties in `Intl` objects (e.g. `Intl.Collator.prototype.compare`)
- New locale aware functions in standard Javascript object prototypes (e.g. `String.prototype.localeCompare`)

> One popular implementation strategy followed by other engines, is to bundle an internationalization framework (typically [ICU](http://site.icu-project.org/)) along with the application package. This guarantees deterministic behaviours at the cost of applications package bloat. We decided to consume the Android platform provided facilities for space efficiency, but at the cost of some variance in behaviours across Android platforms.

Reviewer 2: Love the new contents 👍

- [`Intl.PluralRules`](https://tc39.es/ecma402/#pluralrules-objects)

- [`Intl.RelativeTimeFormat`](https://tc39.es/ecma402/#relativetimeformat-objects)

- [`Intl.DisplayNames`](https://tc39.es/proposal-intl-displaynames/#sec-intl-displaynames-constructor)

- [`Intl.ListFormat`](https://tc39.es/proposal-intl-list-format/#sec-intl-listformat-constructor)

- [`Intl.Locale`](https://tc39.es/ecma402/#sec-intl-locale-constructor)

- `Intl.DateTimeFormat` properties
   - [`dateStyle/timeStyle`](https://tc39.es/proposal-intl-datetime-style/)
   - [`dayPeriod`](https://github.com/tc39/ecma402/issues/29)
   - [`fractionalSecondDigits`](https://github.com/tc39/ecma402/pull/347)
- [`BigInt.prototype.toLocaleString`](https://tc39.es/ecma402/#sup-bigint.prototype.tolocalestring)

## Excluded

- `Intl.DateTimeFormat`: [`formatMatcher`](https://tc39.es/ecma402/#sec-basicformatmatcher) parameter is not respected. The parameter enables the implementation to pick the best display format when it supports only a subset of all possible formats. ICU library in Android platform and hence our implementation allows all subsets and formats which makes this `formatMatcher` property unnecessary.

> ## Known Issues

Reviewer 2: ```suggestion
## Known Issues across Android SDKs
```

Or maybe "Variance across Android SDKs" or "Limitations across Android SDKs" since we are not going to fix them (because of the platform limitations), right?

Author: Makese sense ..

### Android 11

- The keys of the object returned by `resolvedOptions` function in all `Intl` services are not deterministically ordered as prescribed by spec.
- DateFormat: ECMAScript [beginning of time](https://www.ecma-international.org/ecma-262/11.0/index.html#sec-time-values-and-time-range) (-8,640,000,000,000,000), is formatted as `November 271817`, instead of expected `April 271822`.
- `Intl.NumberFormat` implementation has some rough edges in supporting the following properties,
  - `style`: 'unit'
  - `notation`: 'compact'
  - `signDisplay`
  - `currencyFormat`

### Android 10 and older (SDK < 30)

> - `Intl.NumberFormat`: Scientific notation formatting has issues on some cases such as Infinity (e.g. Expected SameValue(«-∞E0», «-∞») to be true).

Reviewer 2: ```suggestion
- `Intl.NumberFormat`: Scientific notation formatting has issues on some cases such as Infinity (e.g. Expected `SameValue(«-∞E0», «-∞»)` to be true).
```

I don't know...are they code?

> - `Intl.NumberFormat`: Compact notation `formatToParts` doesn't identify unit, hence we report unit as 'literal' (Compact short: 987654321: parts[1].type Expected SameValue(«literal», «compact») to be true)

Reviewer 2: I don't actually understand whatever in the `(Compact short: ...)`. The `SameValue` part looks like the spec text but the `«...»` looks like some console output from other engines? I'm not sure. Could you make this more readable?

Author: I've copied out the assert failure statement as it .. my bad :) I'm making the text better readable.

Also, please note that I tried to document the test code with these failures. For instance take a look at : https://github.com/facebook/hermes/blob/master/android/intltest/java/com/facebook/hermes/test/HermesIntlNumberFormatTest.java

Reviewer 2: Oh, so it's error from test262 (well, test402 I guess). I see.

Since you already explained them in prose. Maybe we can just link to those tests if people are interested in looking at the exact edge cases.

> - `Intl.NumberFormat`: `formatToParts` function doesn't produce expected result with Scientific/Engineering notation and input Infinity (`-Infinity` - engineering: length Expected SameValue(«4», «2») to be true).

Reviewer 2: Same here, what is "`-Infinity` - engineering:"?

Author: This one is strange .. When i was testing a few months back, `-Inifinity` was foamtted as something like `∞E0` on older than platform 30 .. but from a quick test, it doesn't seem to be the case anymore .. May be the icu libraries are getting updated over the air .. I'll keep the entry anyways but make the text clearer.

| Android 7.0 - 7.1 (API levels 24 - 25) | ICU4j 56 ([ref](https://developer.android.com/guide/topics/resources/internationalization))| CLDR 28 | Unicode 8.0 |

### Pre-24 platforms

| Android Platform Version | ICU | Unicode | CLDR
| --- | --- | --- | --- |
| Android 6.0 (API level 23) | ICU4j 55.1 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 27.0.1 | Unicode 7.0 |
| Android 5.0 (API levels 21–22) | ICU4j 53 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 25 | Unicode 6.3 |
| Android 4.4 (API levels 19–20) | ICU4j 51 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 23 | Unicode 6.2 |
| Android 4.3 (API level 18) | ICU4j 50 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 22.1 | Unicode 6.2 |
| Android 4.1 (API levels 16–17) | ICU4j 4.8 ([ref](https://developer.android.com/guide/topics/resources/internationalization)) | CLDR 2.0 | Unicode 6.0 |  
  
<br />

> In summary,

Reviewer 2: Maybe this parts should move to the beginning of the "Known issues" sections?

Author: Xuan.. It makes sense to move it  there .. But as it is a summary of the library support on each platform, it make sense to keep it here as well .. It is a tough one .. Shall i keep it here as it for now being ?

Reviewer 2: It's a "maybe" suggestion so I'm fine whatever you choose ;)

> Platforms >= 24 have much better internationalization support than earlier, as many ICU classes are available as is.

Reviewer 2: Maybe an ordered list?
- `30` blah
- `>= 24` blah
- `21-24` blah
- `< 21` blah
